aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -19,42 +19,105 @@
 import org.apache.iceberg.io.PositionOutputStream;
 import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.base.Predicates;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.io.CountingOutputStream;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.MoreExecutors;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import software.amazon.awssdk.core.sync.RequestBody;
 import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.AbortMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CompleteMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CompletedMultipartUpload;
+import software.amazon.awssdk.services.s3.model.CompletedPart;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest;
 import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
 
 class S3OutputStream extends PositionOutputStream {
   private static final Logger LOG = LoggerFactory.getLogger(S3OutputStream.class);
 
+  static final String UPLOAD_POOL_SIZE  = "s3fileio.multipart.num-threads";
+  static final String MULTIPART_SIZE = "s3fileio.multipart.part.size";
+  static final String MULTIPART_THRESHOLD_FACTOR = "s3fileio.multipart.threshold";
+
+  static final int DEFAULT_UPLOAD_WORKER_POOL_SIZE = Runtime.getRuntime().availableProcessors();
+  static final int MIN_MULTIPART_UPLOAD_SIZE = 5 * 1024 * 1024;
+  static final int DEFAULT_MULTIPART_SIZE = 32 * 1024 * 1024;
+  static final double DEFAULT_MULTIPART_THRESHOLD = 1.5;
+
+  private static ExecutorService executorService;
+
   private final StackTraceElement[] createStack;
   private final S3Client s3;
   private final S3URI location;
 
-  private final OutputStream stream;
-  private final File stagingFile;
-  private long pos = 0;
+  private CountingOutputStream stream;
+  private final List<File> stagingFiles = Lists.newArrayList();
+  private File currentStagingFile;
+  private String multipartUploadId;
+  private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
+  private final int multiPartSize;
+  private final double multiPartThresholdFactor;
 
+  private long pos = 0;
   private boolean closed = false;
 
-  S3OutputStream(S3Client s3, S3URI location) throws IOException {
+  S3OutputStream(Map<String, String> properties, S3Client s3, S3URI location) throws IOException {
+    synchronized (this) {
+      if (executorService == null) {
```

Comment:
If this check were done outside of `synchronized` as well, then we wouldn't need to get the lock each time a file is opened. There isn't a huge chance of lock contention, but it is always nice to avoid locking when possible.

Reply:
I'll add the double-check, but like you said, this really isn't a critical section (and I believe the the stack grab we do below is probably much more expensive and likely to cause some sort of locking).